doomemacs/.config/doom/config.org
#+title:  Doomemacs
#+author:  bladrome
#+property:  header-args  :mkdirp yes


* Install

#+end_src bash :tangle no
emacs --batch doomemacs.org -f org-babel-tangle
#+end_src


* Init

#+begin_src elisp :tangle init.el

;;; init.el -*- lexical-binding: t; -*-

;; This file controls what Doom modules are enabled and what order they load
;; in. Remember to run 'doom sync' after modifying it!

;; NOTE Press 'SPC h d h' (or 'C-h d h' for non-vim users) to access Doom's
;;      documentation. There you'll find a link to Doom's Module Index where all
;;      of our modules are listed, including what flags they support.

;; NOTE Move your cursor over a module's name (or its flags) and press 'K' (or
;;      'C-c c k' for non-vim users) to view its documentation. This works on
;;      flags as well (those symbols that start with a plus).
;;
;;      Alternatively, press 'gd' (or 'C-c c d') on a module to browse its
;;      directory (for easy access to its source code).

(doom! :input
       ;;bidi              ; (tfel ot) thgir etirw uoy gnipleh
       ;;chinese
       ;;japanese
       ;;layout            ; auie,ctsrnm is the superior home row

       :completion
       ;;company           ; the ultimate code completion backend
       (corfu
        +orderless)        ; complete with cap(f), cape and a flying feather!
       ;;helm              ; the *other* search engine for love and life
       ;;ido               ; the other *other* search engine...
       ;; (ivy
       ;;  +fuzzy
       ;;  +prescient
       ;;  +icons)            ; a search engine for love and life
       (vertico
        :icons
        :childframe)

       :ui
       ;;deft              ; notational velocity for Emacs
       doom                ; what makes DOOM look the way it does
       doom-dashboard      ; a nifty splash screen for Emacs
       ;;doom-quit         ; DOOM quit-message prompts when you quit Emacs
       ;;(emoji +unicode)  ; 🙂
       hl-todo             ; highlight TODO/FIXME/NOTE/DEPRECATED/HACK/REVIEW
       ;;indent-guides     ; highlighted indent columns
       (ligatures +extra)  ; ligatures and symbols to make your code pretty again
       ;;minimap           ; show a map of the code on the side
       modeline            ; snazzy, Atom-inspired modeline, plus API
       ;;nav-flash         ; blink cursor line after big motions
       ;;neotree           ; a project drawer, like NERDTree for vim
       ophints             ; highlight the region an operation acts on
       (popup
        +all
        +defaults)         ; tame sudden yet inevitable temporary windows
       ;;tabs              ; a tab bar for Emacs
       treemacs            ; a project drawer, like neotree but cooler
       unicode             ; extended unicode support for various languages
       (vc-gutter
        +pretty)           ; vcs diff in the fringe
       vi-tilde-fringe     ; fringe tildes to mark beyond EOB
       (window-select
        +switch-window)    ; visually switch windows
       workspaces          ; tab emulation, persistence & separate workspaces
       zen                 ; distraction-free coding or writing

       :editor
       ;;evil              ; come to the dark side, we have cookies
       file-templates      ; auto-snippets for empty files
       fold                ; (nigh) universal code folding
       format              ; automated prettiness
       ;;god               ; run Emacs commands without modifier keys
       ;;lispy             ; vim for lisp, for people who don't like vim
       multiple-cursors    ; editing in many places at once
       ;;objed             ; text object editing for the innocent
       ;;parinfer          ; turn lisp into python, sort of
       rotate-text         ; cycle region at point between text candidates
       snippets            ; my elves. They type so I don't have to
       word-wrap           ; soft wrapping with language-aware indent

       :emacs
       (dired
        +dirvish
        +icons)            ; making dired pretty [functional]
       electric            ; smarter, keyword-based electric-indent
       eww                 ; the internet is gross
       (ibuffer +icons)    ; interactive buffer management
       undo                ; persistent, smarter undo for your inevitable mistakes
       vc                  ; version-control and Emacs, sitting in a tree

       :term
       eshell              ; the elisp shell that works everywhere
       ;;shell             ; simple shell REPL for Emacs
       ;;term              ; basic terminal emulator for Emacs
       ;;vterm             ; the best terminal emulation in Emacs

       :checkers
       syntax              ; tasing you for every semicolon you forget
       (spell +enchant)    ; tasing you for misspelling mispelling
       ;; grammar             ; tasing grammar mistake every you make

       :tools
       ;;ansible
       biblio              ; Writes a PhD for you (citation needed)
       collab              ; buffers with friends
       (debugger
        +lsp)              ; FIXME stepping through code, to help you add bugs
       ;;direnv
       docker
       ;;editorconfig      ; let someone else argue about tabs vs spaces
       ;;ein               ; tame Jupyter notebooks with emacs
       (eval +overlay)     ; run code, run (also, repls)
       (lookup
        +dictionary
        +offline)          ; navigate your code and its documentation
       (lsp +peek)         ; M-x vscode
       magit               ; a git porcelain for Emacs
       make                ; run make tasks from Emacs
       pass                ; password manager for nerds
       pdf                 ; pdf enhancements
       ;;prodigy           ; FIXME managing external services & code builders
       ;;terraform         ; infrastructure as code
       tmux                ; an API for interacting with tmux
       tree-sitter         ; syntax and parsing, sitting in a tree...
       upload              ; map local to remote projects via ssh/ftp

       :os
       (:if (featurep :system 'macos) macos)  ; improve compatibility with macOS
       tty                 ; improve the terminal Emacs experience

       :lang

       ;;agda              ; types of types of types of types...
       beancount           ; mind the GAAP
       (cc +lsp)           ; C > C++ == 1
       ;;clojure           ; java with a lisp
       common-lisp         ; if you've seen one lisp, you've seen them all
       ;;coq               ; proofs-as-programs
       ;;crystal           ; ruby at the speed of c
       ;;csharp            ; unity, .NET, and mono shenanigans
       data                ; config/data formats
       ;;(dart +flutter)   ; paint ui and not much else
       ;;dhall
       ;;elixir            ; erlang done right
       ;;elm               ; care for a cup of TEA?
       emacs-lisp          ; drown in parentheses
       ;;erlang            ; an elegant language for a more civilized age
       ess                 ; emacs speaks statistics
       ;;factor
       ;;faust             ; dsp, but you get to keep your soul
       ;;fortran           ; in FORTRAN, GOD is REAL (unless declared INTEGER)
       ;;fsharp            ; ML stands for Microsoft's Language
       ;;fstar             ; (dependent) types and (monadic) effects and Z3
       ;;gdscript          ; the language you waited for
       (go +lsp)           ; the hipster dialect
       ;;(graphql +lsp)    ; Give queries a REST
       ;;(haskell +lsp)    ; a language that's lazier than I am
       ;;hy                ; readability of scheme w/ speed of python
       ;;idris             ; a language you can depend on
       json                ; At least it ain't XML
       ;;(java +lsp)       ; the poster child for carpal tunnel syndrome
       (javascript +lsp)   ; all(hope(abandon(ye(who(enter(here))))))
       julia               ; a better, faster MATLAB
       ;;kotlin            ; a better, slicker Java(Script)
       (latex
        +cdlatex
        +latexmk
        +lsp)              ; writing papers in Emacs has never been so fun
       ;;lean              ; for folks with too much to prove
       ;;ledger            ; be audit you can be
       (lua +lsp)          ; one-based indices? one-based indices
       markdown            ; writing docs for people to ignore
       ;;nim               ; python + lisp at the speed of c
       ;;nix               ; I hereby declare "nix geht mehr!"
       ;;ocaml             ; an objective camel
       (org
        +roam2
        +pandoc
        +dragndrop
        +gnuplot
        +hugo
        +journal
        +noter
        +pretty
        +present)          ; organize your plain life in plain text
       ;;php               ; perl's insecure younger brother
       plantuml            ; diagrams for confusing people more
       graphviz            ; diagrams for confusing yourself even more
       ;;purescript        ; javascript, but functional
       (python
        +anaconda
        +conda
        +poetry
        +lsp
        +tree-sitter
        +pyright)          ; beautiful is better than ugly
       ;;qt                ; the 'cutest' gui framework ever
       ;;racket            ; a DSL for DSLs
       ;;raku              ; the artist formerly known as perl6
       rest                ; Emacs as a REST client
       ;;rst               ; ReST in peace
       ;;(ruby +rails)     ; 1.step {|i| p "Ruby is #{i.even? ? 'love' : 'life'}"}
       (rust +lsp)         ; Fe2O3.unwrap().unwrap().unwrap().unwrap()
       ;;scala             ; java, but good
       ;;(scheme +guile)   ; a fully conniving family of lisps
       sh                  ; she sells {ba,z,fi}sh shells on the C xor
       ;;sml
       ;;solidity          ; do you need a blockchain? No.
       ;;swift             ; who asked for emoji variables?
       ;;terra             ; Earth and Moon in alignment for performance.
       web                 ; the tubes
       yaml                ; JSON, but readable
       zig                 ; C, but simpler

       :email
       (mu4e +org)
       ;;notmuch
       ;;(wanderlust +gmail)

       :app
       calendar
       emms
       everywhere          ; *leave* Emacs!? You must be joking
       ;;irc               ; how neckbeards socialize
       (rss +org)          ; emacs as an RSS reader

       :config
       literate
       (default
         +snippets
         +bindings
         +smartparens))

#+end_src


* Config
** Base Config

#+begin_src elisp :tangle config.el

;; (setq toggle-debug-on-error t)

(setq user-full-name "bladrome"
      user-mail-address "blackwhitedoggie@163.com")

(use-package! package
  :config
  (setq package-archives
        '(("melpa" . "http://mirrors.bfsu.edu.cn/elpa/melpa/")
          ("org"   . "http://mirrors.bfsu.edu.cn/elpa/org/")
          ("gnu"   . "http://mirrors.bfsu.edu.cn/elpa/gnu/")))
  (package-initialize))

(use-package! use-package
  :custom
  (use-package-always-ensure nil)
  (package-native-compile t)
  (warning-minimum-level :emergency))


(setq-default cursor-type 'box)
(setq browse-url-browser-function 'browse-url-xdg-open
      browse-url-generic-program 'browse-url-xdg-open)


(global-visual-line-mode)

;; t relative or nil
(setq display-line-numbers-type nil)

(show-paren-mode t)

(setq use-short-answers t)

(blink-cursor-mode 1)

(setq-default x-stretch-cursor t
              x-underline-at-descent-line t)


(setq shell-file-name (executable-find "bash"))

(setq confirm-kill-emacs nil
      system-time-locale "C"
      pop-up-windows t
      scroll-margin 2
      widget-image-enable nil)

(setq byte-compile-warnings '(not obsolete))
(setq warning-suppress-log-types '((comp) (bytecomp)))
(setq native-comp-async-report-warnings-errors 'silent)
(setq inhibit-startup-echo-area-message (user-login-name))
(setq visible-bell t)
(setq ring-bell-function 'ignore)
(setq set-message-beep 'silent)

;; coding
(prefer-coding-system 'utf-8)
(set-default-coding-systems 'utf-8)
(setq default-buffer-file-coding-system 'utf-8)


(setq delete-by-moving-to-trash t)

(setq word-wrap-by-category t)

(save-place-mode 1)

(global-auto-revert-mode)

(setq initial-major-mode 'org-mode)
(setq initial-scratch-message nil)

;; Smooth mouse scrolling
(setq mouse-wheel-scroll-amount '(2 ((shift) . 1))
      mouse-wheel-progressive-speed nil
      mouse-wheel-follow-mouse t
      scroll-step 1)

(setq fancy-splash-image (file-name-concat doom-user-dir "splash.png"))
#+end_src

*** bookmark

#+begin_src elisp :tangle config.el
;; bookmarks and recentf cache

(after! bookmark
  (setq bookmark-default-file (file-name-concat doom-user-dir "bookmarks")))

(after! recentf
  (setq recentf-save-file (file-name-concat doom-user-dir "recentf")))

#+end_src

*** proxy

#+begin_src elisp :tangle config.el
;; (setq url-proxy-services
;;          '(("socks5" . "localhost:8888")))
#+end_src

*** auth

#+begin_src elisp :tangle config.el
(setq auth-sources (list (file-name-concat doom-user-dir "authinfo.gpg")
                         "~/.authinfo.gpg"))
#+end_src

*** transparency

#+begin_src elisp :tangle config.el

;; Set transparency of emacs
(defun transparency (value)
  "Sets the transparency of the frame window. 0=transparent/100=opaque"
  (interactive "nTransparency Value 0 - 100 opaque:")
  (set-frame-parameter (selected-frame) 'alpha (if (> value 100) 100 value)))

(transparency 100)
#+end_src

*** search completion


#+begin_src elisp :tangle config.el

;; MINIBUFFER COMPLETION
(use-package! vertico
  :init (vertico-mode)
  :custom
  (vertico-sort-function 'vertico-sort-history-alpha))

;; Search for partial matches in any order
(use-package! orderless
  :custom
  (completion-styles '(orderless basic))
  (completion-category-defaults nil)
  (completion-category-overrides
   '((file (styles partial-completion)))))

;; Improve keyboard shortcut discoverability
(use-package! which-key
  :config (which-key-mode)
  :custom
  (which-key-max-description-length 40)
  (which-key-lighter nil)
  (which-key-sort-order 'which-key-description-order))

;; Consult convenience functions
(use-package! consult
  :bind
  (("C-c M-x" . consult-mode-command)
   ("C-c h" . consult-history)
   ("C-c k" . consult-kmacro)
   ("C-c m" . consult-man)
   ("C-c i" . consult-info)
   ([remap Info-search] . consult-info)))

;; Improved help buffers
(use-package! helpful
  :bind
  (("C-h f" . helpful-function)
   ("C-h x" . helpful-command)
   ("C-h k" . helpful-key)
   ("C-h v" . helpful-variable)))
#+end_src

** Theme

#+begin_src elisp :tangle packages.el
(package! nova
  :recipe (:host github :repo "thisisran/nova"))

(package!
  svg-tag-mode
  :recipe (:host github :repo "rougier/svg-tag-mode"))

(package! nano-theme
  :recipe (:host github :repo "rougier/nano-theme"))

(package! nano-modeline)

(package! lambda-themes
  :recipe (:host github :repo "Lambda-Emacs/lambda-themes"))

(package! ef-themes
  :recipe (:host github :repo "protesilaos/ef-themes"))

#+end_src

#+begin_src elisp :tangle config.el
(use-package! nova
  :config
  (setq nova-vertico-mode 1))

(setq font-size
      (if (string= (system-name) "penguin")
          40
        20))
(setq doom-font (font-spec :family "SauceCodePro Nerd Font Mono" :size font-size)
      doom-variable-pitch-font doom-font
      doom-symbol-font (font-spec :family "VictorMono")
      doom-big-font (font-spec :family "VictorMono" :size (+ font-size 6)))

(setq native-comp-deferred-compilation nil)
(setq native-comp-jit-compilation nil)


(add-hook! 'after-setting-font-hook
  (set-fontset-font t 'symbol (font-spec :family "Symbola"))
  (set-fontset-font t 'mathematical (font-spec :family "Symbola"))
  (set-fontset-font t 'emoji (font-spec :family "Symbola")))

#+end_src

Select theme.

#+begin_src elisp :tangle config.el

(use-package! doom-modeline
  :custom
  (doom-modeline-buffer-encoding t)
  (doom-modeline-enable-word-count nil)
  (doom-modeline-height 10))

;; (setq +doom-dashboard-functions '(doom-dashboard-widget-banner))
;; (setq doom-theme 'lambda-light-faded)
(setq doom-theme 'doom-one)

;; (setq doom-theme nil)
;; (require 'nano-theme)
;; (nano-mode)
;; (nano-light)
;; (setq-default mode-line-format nil)
;; (use-package! nano-modeline)
;; (add-hook 'org-mode-hook  #'nano-modeline-org-mode)
#+end_src

** Org

#+begin_src elisp :tangle packages.el

(package! org-page)
(package! org-ql)
(package! org-special-block-extras)

(package! org-modern)
(package! org-download)
(package! org-web-tools)

(package! org-graph-view
  :recipe (:host github :repo "alphapapa/org-graph-view"))

(package! org-pandoc-import
  :recipe (:host github
           :repo "tecosaur/org-pandoc-import"
           :files ("*.el" "filters" "preprocessors")))

(package! org-transclusion
  :recipe (:host github :repo "nobiot/org-transclusion"))

(package! org-tufte
  :recipe (:host github :repo "Zilong-Li/org-tufte"))

(package! org-protocol-capture-html
    :recipe (:host github :repo "alphapapa/org-protocol-capture-html"))

#+end_src

#+begin_src elisp :tangle config.el
;; (after! conda
;;   (setq conda-anaconda-home "/run/media/bladrome/bank/drome/mambaforge"
;; 	conda-env-home-directory  "/run/media/bladrome/bank/drome/mambaforge"))

(after! org
  (setq org-tag-alist (quote (("noexport" . ?n)
                              (:startgroup)
                              ("@office" . ?o)
                              ("@field" . ?f)
                              (:endgroup)
                              ("personal" . ?p)
                              ("work" . ?w)
                              ("cancelled" . ?c)
                              ("read" . ?r)
                              ("browse" . ?b)
                              ("flagged" . ??))))
  ;; Allow setting single tags without the menu
  (setq org-fast-tag-selection-single-key (quote expert))
  ;; For tag searches ignore tasks with scheduled and deadline dates
  (setq org-agenda-tags-todo-honor-ignore-options t)
  (setq org-startup-folded "folded")
  (setq org-export-date-timestamp-format nil)
  ;; org-directory
  (setq org-directory "~/Documents/Org")
  (setq org-hugo-base-dir (file-name-concat org-directory "Hugo"))
  (setq org-agenda-files (list (file-name-concat org-directory "gtd")))
  (setq org-log-done 'time)
  (setq org-log-redeadline 'time)
  (setq org-log-reschedule 'time)
  (setq org-agenda-hide-tags-regexp ".")
  (setq org-refile-allow-creating-parent-nodes 'confirm)
  (setq org-attach-id-dir (file-name-concat org-directory "attachments")))




(use-package! org-web-tools
  :init
  (map!
   :leader
   :prefix "i"
   :desc "insert weblinks" "l" #'org-web-tools-insert-link-for-url))

(use-package! org-download
  :defer nil
  :custom
  (org-download-image-dir (file-name-concat org-attach-id-dir (format-time-string "%Y%m%d")))
  (org-image-actual-width '(600))
  (org-download-heading-lvl nil)
  (org-download-timestamp "")
  :config
  (require 'org-download))

(after! org-download
  (setq org-download-method 'directory)
  (setq org-download-link-format "[[file:images/%s]]\n"))

;; org-export docs using a tempalate.docx
(defun org-export-docx ()
  (interactive)
  (let ((docx-file (concat (file-name-sans-extension (buffer-file-name)) ".docx"))
	(template-file (file-name-concat doom-user-dir "template.docx")))
    (shell-command (format "pandoc %s -o %s --reference-doc=%s" (buffer-file-name) docx-file template-file))
    ;; pandoc --bibliography=/bibliographyfile.bib \
    ;;        --csl=chicago-fullnote-bibliography.csl \
    ;;        --reference-docx=reference.docx \
    ;;        -i (buffer-file-name) -o docx.docx
    (message "Convert finish: %s" docx-file)))


(after! org
  (setq org-confirm-babel-evaluate nil
        org-edit-src-content-indentation 0
        org-fold-catch-invisible-edits 'show
        org-fontify-quote-and-verse-blocks t
        org-fontify-whole-heading-line t
        org-fontify-done-headline t
        org-hide-block-startup t
        org-hide-emphasis-markers t
        org-hide-leading-stars t
        org-id-link-to-org-use-id t
        org-image-actual-width '(450)
        org-list-allow-alphabetical t
        org-pretty-entities t
        org-return-follows-link t
        org-special-ctrl-a/e t
        org-special-ctrl-k t
        org-src-preserve-indentation nil
        org-src-tab-acts-natively t
        org-startup-indented t
        org-startup-with-inline-images t
        org-use-speed-commands t
        org-use-sub-superscripts nil
        org-startup-with-inline-images nil))

;; (add-hook! 'org-babel-after-execute-hook 'org-display-inline-images 'append)
;; (add-hook! 'org-babel-after-execute-hook 'org-toggle-latex-fragment 'append)

(use-package! org-special-block-extras
  :hook (org-mode . org-special-block-extras-mode))

(use-package! svg-lib)
(use-package! svg-tag-mode)

(use-package! org
  :hook
  (org-mode . turn-on-visual-line-mode)
  :config
  (pdf-loader-install))



(after! beancount
  (setq beancount-number-alignment-column 59))
#+end_src

** Latex
*** LaTeX Export

#+begin_src elisp :tangle packages.el
(package! org-fragtog)
#+end_src

#+begin_src elisp :tangle config.el

;; org-latex-compilers = ("pdflatex" "xelatex" "lualatex"), which are the possible values for %latex
(setq org-latex-pdf-process '("LC_ALL=en_US.UTF-8 latexmk -f -pdf -%latex -shell-escape -interaction=nonstopmode -output-directory=latex.out %f"))

(setq-default org-html-with-latex `dvisvgm)

(after! ox-latex
  (add-to-list 'org-latex-classes
	       '("elsarticle"
		 "\\documentclass[review,authoryear]{elsarticle}"
		 ("\\section{%s}" . "\\section*{%s}")
		 ("\\subsection{%s}" . "\\subsection*{%s}")
		 ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
		 ("\\paragraph{%s}" . "\\paragraph*{%s}")
		 ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))))

(setq Tex-command-default "XeLaTeX")
(setq TeX-engine 'xelatex)
(setq TeX-command-extra-options "-shell-escape")

(use-package! org-fragtog
  :after org
  :custom
  (org-startup-with-latex-preview t)
  :hook
  (org-mode . org-fragtog-mode)
  :custom
  (org-format-latex-options
   (plist-put org-format-latex-options :scale (/ (frame-char-height) 10))
   (plist-put org-format-latex-options :foreground 'auto)
   (plist-put org-format-latex-options :background 'auto)))

(use-package! org
  :config
  (add-to-list 'org-latex-packages-alist '("outputdir=latex.out" "minted"))
  (setq org-latex-minted-options
        '(("bgcolor" . "bg")
          ("breaklines" . "true")
          ("autogobble" . "true")
          ("fontsize" . "\\scriptsize")))
  (setq org-latex-listings 'minted))

#+end_src

*** LaTeX Fragments

**** Prettier highlighting

#+begin_src elisp :tangle config.el
(setq org-highlight-latex-and-related '(latex script entities))
(require 'org-src)
(add-to-list 'org-src-block-faces '("latex" (:inherit default :extend t)))
#+end_src

** Dired

#+begin_src elisp :tangle config.el

(after! dired
  (setq dired-listing-switches "-aBhl  --group-directories-first"
	dired-dwim-target t
	dired-recursive-copies (quote always)
	dired-recursive-deletes (quote top)
	;; Directly edit permisison bits!
	wdired-allow-to-change-permissions t
	dired-omit-mode nil))

#+end_src

** Agenda

#+begin_src elisp :tangle packages.el
(package! org-super-agenda)
#+end_src

#+begin_src elisp :tangle config.el
(use-package! org-super-agenda
  :commands org-super-agenda-mode)

(after! org-agenda
  (let ((inhibit-message t))
    (org-super-agenda-mode)))

(setq org-agenda-skip-scheduled-if-done t
      org-agenda-skip-deadline-if-done t
      org-agenda-include-deadlines t
      org-agenda-block-separator nil
      org-agenda-tags-column 100 ;; from testing this seems to be a good value
      org-agenda-compact-blocks t)

(setq org-agenda-custom-commands
      '(("o" "Overview"
         ((agenda "" ((org-agenda-span 'day)
                      (org-super-agenda-groups
                       '((:name "Today"
                          :time-grid t
                          :date today
                          :todo "TODAY"
                          :scheduled today
                          :order 1)))))
          (alltodo "" ((org-agenda-overriding-header "")
                       (org-super-agenda-groups
                        '((:name "Next to do"
                           :todo "NEXT"
                           :order 1)
                          (:name "Important"
                           :tag "Important"
                           :priority "A"
                           :order 6)

doomemacs/.config/doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!

;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "bladrome"
      user-mail-address "blkcat@163.com")

(setq package-archives
    '(("melpa" . "http://mirrors.bfsu.edu.cn/elpa/melpa/")
      ("org"   . "http://mirrors.bfsu.edu.cn/elpa/org/")
      ("gnu"   . "http://mirrors.bfsu.edu.cn/elpa/gnu/")))
;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 19 :weight 'semi-light) ;
      ;; doom-variable-pitch-font (font-spec :family "DejaVu Sans" :size 20))

(setq doom-font (font-spec :family "Fira Mono" :size 19)
     doom-serif-font (font-spec :family "Fira Mono" :size 20)
     ;; doom-variable-pitch-font (font-spec :family "SourceHanSerifCN")
     doom-unicode-font (font-spec :family "SourceHanSerifCN")
     ;; doom-big-font (font-spec :family "SourceHanSerifCN" :size 23)
     )



;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-ayu-mirage)
;; (setq doom-theme nil)
;; (require 'disp-table)
;; (require 'nano-faces)
;; (require 'nano-colors)
;; (require 'nano-theme)
;; (require 'nano-theme)
;; ;; (require 'nano-help)
;; ;; (require 'nano-splash)
;; ;; (require 'nano-modeline)
;; (nano-faces)
;; (nano-theme)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")
(setq org-attach-id-dir "/home/bladrome/Documents/2021/attachments")
;; (setq org-latex-pdf-process (list "latexmk -shell-escape -bibtex -f -pdfxe %f"))


;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

(setq bookmark-default-file "~/.config/doom/bookmarks")
(setq recentf-save-file "~/.config/doom/recentf")

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.
;; (setq debug-on-quit t)
;; (setq debug-on-error t)
;; set transparency
;; (set-frame-parameter (selected-frame) 'alpha '(95 85))
;; (add-to-list 'default-frame-alist '(alpha 95 85))

(setq org-export-babel-evaluate nil)
(setq org-publish-project-alist
      '(("orgfiles"
         :base-directory "~/Documents/2021/"
         :base-extension "org"
         :publishing-directory "~/org/public_html"
         :publishing-function org-html-publish-to-html
         :with-toc t
         :auto-preamble t
         :auto-sitemap
         :sitemap-title "Notes"
         :sitemap-sort-files
         :html-head "<link rel=\"stylesheet\" type=\"text/css\" href=\"style/worg.css\" />"
         :html-preamble nil)
        ("images"
         :base-directory "~/Documents/2021/attachments"
         :base-extension "png\\|jpg\\|webp"
         :recursive t
         :publishing-directory "~/org/public_html/images"
         :publishing-function org-publish-attachment)
        ("other"
         :base-directory "~/other/"
         :base-extension "css\\|el"
         :publishing-directory "~/org/public_html/others"
         :recursive t
         :publishing-function org-publish-attachment)
        ("org" :components ("orgfiles" "images" "other"))))


(after! dired
  (setq dired-listing-switches "-aBhl  --group-directories-first"
        dired-dwim-target t
        dired-recursive-copies (quote always)
        dired-recursive-deletes (quote top)
        ;; Directly edit permisison bits!
        wdired-allow-to-change-permissions t
        dired-omit-mode nil))


(use-package! org-special-block-extras
  :ensure t
  :hook (org-mode . org-special-block-extras-mode)
  ;; All relevant Lisp functions are prefixed ‘o-’; e.g., `o-docs-insert'.
  ;; :config
  ;; (o-defblock src (lang nil) (title nil exports nil file nil)
  ;; "Fold-away all ‘src’ blocks as ‘<details>’ HTML export.
  ;;       If a block has a ‘:title’, use that to title the ‘<details>’."
  ;; (format "<details> <summary> %s </summary> <pre> %s </pre></details>"
  ;;         (or title (concat "Details; " lang))
  ;;         raw-contents))
)


(use-package! pangu-spacing
  :config
  (global-pangu-spacing-mode 1)
  (setq pangu-spacing-real-insert-separtor t))

(use-package! valign
  :config
  (setq valign-fancy-bar t)
  (add-hook 'org-mode-hook #'valign-mode))

(use-package leetcode
    :config
    (setq leetcode-save-solutions t
          leetcode-prefer-language "cpp"
          leetcode-prefer-sql "mysql"
          leetcode-directory "~/workground/Leetcode/"))


(use-package org-ref
  :config
  (setq bibtex-completion-bibliography '("~/Documents/2021/bibliography/bibliography.bib"
                                         "~/Documents/2021/bibliography/arxiv.bib"
                                         "~/Documents/2021/bibliography/references.bib")
        bibtex-completion-library-path '("~/Dropbox/emacs/bibliography/bibtex-pdfs/")
        bibtex-completion-notes-path "~/Dropbox/emacs/bibliography/notes/"
        bibtex-completion-notes-template-multiple-files "* ${author-or-editor}, ${title}, ${journal}, (${year}) :${=type=}: \n\nSee [[cite:&${=key=}]]\n"
        bibtex-completion-additional-search-fields '(keywords)
        bibtex-completion-display-formats
        '((article       . "${=has-pdf=:1}${=has-note=:1} ${year:4} ${author:36} ${title:*} ${journal:40}")
          (inbook        . "${=has-pdf=:1}${=has-note=:1} ${year:4} ${author:36} ${title:*} Chapter ${chapter:32}")
          (incollection  . "${=has-pdf=:1}${=has-note=:1} ${year:4} ${author:36} ${title:*} ${booktitle:40}")
          (inproceedings . "${=has-pdf=:1}${=has-note=:1} ${year:4} ${author:36} ${title:*} ${booktitle:40}")
          (t             . "${=has-pdf=:1}${=has-note=:1} ${year:4} ${author:36} ${title:*}"))
        bibtex-completion-pdf-open-function
        (lambda (fpath)
          (call-process "open" nil 0 nil fpath)))
  (require 'bibtex)
  (setq bibtex-autokey-year-length 4
        bibtex-autokey-name-year-separator "-"
        bibtex-autokey-year-title-separator "-"
        bibtex-autokey-titleword-separator "-"
        bibtex-autokey-titlewords 2
        bibtex-autokey-titlewords-stretch 1
        bibtex-autokey-titleword-length 5
        org-ref-bibtex-hydra-key-binding (kbd "H-b"))
  (define-key bibtex-mode-map (kbd "H-b") 'org-ref-bibtex-hydra/body)
  (require 'org-ref)
  (setq reftex-default-bibliography '("~/Documents/2021/bibliography/references.bib"))
  (setq org-ref-bibliography-notes "~/Documents/2021/bibliography/notes.org"
        org-ref-default-bibliography '("~/Documents/2021/bibliography/references.bib")
        org-ref-pdf-directory "~/Documents/2021/bibliography/bibtex-pdfs/")
    ;; (require 'helm-bibtex)
  ;; (require 'org-ref-helm)
  ;; (setq org-ref-insert-link-function 'org-ref-insert-link-hydra/body
  ;;       org-ref-insert-cite-function 'org-ref-cite-insert-helm
  ;;       org-ref-insert-label-function 'org-ref-insert-label-link
  ;;       org-ref-insert-ref-function 'org-ref-insert-ref-link
  ;;       org-ref-cite-onclick-function (lambda (_) (org-ref-citation-hydra/body)))
    )

;; (server-start)
(use-package org
  :hook
  (org-mode . turn-on-visual-line-mode)
  :config
  ; Tags with fast selection keys
  (setq org-tag-alist (quote ((:startgroup)
                              ("@office" . ?o)
                              ("@field" . ?f)
                              (:endgroup)
                              ("personal" . ?p)
                              ("work" . ?w)
                              ("cancelled" . ?c)
                              ("read" . ?r)
                              ("browse" . ?b)
                              ("flagged" . ??))))
  ; Allow setting single tags without the menu
  (setq org-fast-tag-selection-single-key (quote expert))
  ; For tag searches ignore tasks with scheduled and deadline dates
  (setq org-agenda-tags-todo-honor-ignore-options t)
  ;; (require 'org-bars)
  ;; (add-hook 'org-mode-hook #'org-bars-mode)
  (setq org-startup-folded "folded")

  ;; Capture templates for links to pages having [ and ]
  ;; characters in their page titles - notably ArXiv
  ;; From https://github.com/sprig/org-capture-extension
  (require 'org-protocol)
  (defun transform-square-brackets-to-round-ones(string-to-transform)
    "Transforms [ into ( and ] into ), other chars left unchanged."
    (concat
     (mapcar #'(lambda (c) (if (equal c ?\[) ?\( (if (equal c ?\]) ?\) c))) string-to-transform)))
  (setq org-capture-templates `(
                                ("p" "Protocal" entry (file+headline ,(concat "~/Documents/2021/" (format-time-string "%Y%m%d") ".org") "arxiv")
                                 "* [[%:link][%(transform-square-brackets-to-round-ones \"%:description\")]]\n \n%i\n\n\n\n%?")
                                ("L" "Protocol Link" entry (file+headline ,(concat org-directory "notes_" (shell-command-to-string "date +%F__%H-%M-%S_%Z")) "Inbox")
                                 "* %^{Title_and_tag}\n [[%:link][%(transform-square-brackets-to-round-ones \"%:description\")]]\n")
                                ("w" "Web site" entry (file+headline ,(concat "~/Documents/2021/" (format-time-string "%Y%m%d") ".org") "arxiv")
                                 "* %a :website:\n\n%U %?\n\n%:initial")
                                ("c" "Captured" entry (file+headline ,(concat "~/Documents/2021/" (format-time-string "%Y%m%d") ".org") "arxiv")
                                 "* %t %:description\nlink: %l \n\n%i\n" :prepend t :empty-lines-after 1)
                                ("n" "Captured Now!" entry (file+headline ,(concat "~/Documents/2021/" (format-time-string "%Y%m%d") ".org") "arxiv")
                                 "* %t %:description\nlink: %l \n\n%i\n" :prepend t :emptry-lines-after 1 :immediate-finish t)
                                )))

;; "* %^{Title_and_tag}\nSource: [[%:link][%(transform-square-brackets-to-round-ones \"%:description\")]]\n#+BEGIN_QUOTE\n%i\n#+END_QUOTE\n\n\n%?")
